### PR TITLE
New feature/ids field finder

### PR DESCRIPTION
**Pull request by mgyoo86**
## Summary
This PR introduces a new `findall` method that searches for specific fields within multiple `IDS` objects.
The target fields can be specified using either Symbols or Regex patterns.
The `findall` method returns an IDS_Field_Finder struct for each matched field, containing detailed information that allows for verbose terminal output and plot recipes.

Note: The plot `@recipes` will be represnted in another PR soon.

## Examples
Here are some quick examples.

### 1. Default behavior
```julia
# By default, it searches everything under the given IDS
# Following commands produce the same result
findall(dd.equilibrium.time_slice[].global_quantities)
findall(dd.equilibrium.time_slice[].global_quantities, r"") # Same behavior (Default)
findall(dd.equilibrium.time_slice[].global_quantities, :all) # Same behavior
```

### 2. Finding a single field within a single root IDS
```julia
 # Find fields matching a single symbol within a IDS structure
IFF = findall(dd.equilibrium.time_slice, :psi)
```

### 3. Finding multiple fields within multiple root IDS objects
```julia
 # Search for multiple symbols within multiple root IDS objects
IFF = findall([dd.equilibrium, dd.core_sources], [:psi, :j_tor])
```

### 4. Keyword argument of `include_subfields::Bool=true` 
```julia
  # Control subfield inclusion using the `include_subfields` keyword
IFF = findall(dd, r"prof.*2d"; include_subfields=false)
IFF = findall(dd, r"prof.*2d"; include_subfields=true) # Default behavior
```

### 5. Lazy evaluation of the value of found field objects

```julia
 # Retrieve actual values of found field objects (lazy evaluation)
IFF[1].value
IFF[end].value
```

### CODE REVIEW

File: src/data.jl [hunk 1]
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
...
+end
+
+function Base.show(io::IO, ::MIME"text/plain", IFF::IDS_Field_Finder)
+    root_name = location(IFF.root_ids)
+    parent_name = location(IFF.parent_ids)
+
+    rest_part = replace(parent_name, root_name => "")
+
+    printstyled(io, root_name; color=:red)
+    if IFF.root_ids isa IMASdd.dd || root_name == parent_name
```
Comment (orso82) [inline src/data.jl:904]:
```suggestion
    if IFF.root_ids isa DD || root_name == parent_name
```
or
```suggestion
    if typeof(IFF.root_ids) <: DD || root_name == parent_name
```

File: src/data.jl [hunk 2]
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
...
+        print(io, ".")
+    end
+    isempty(rest_part) ? nothing : print(io, rest_part * ".")
+
+    printstyled(io, String(IFF.field); color=:green, bold=true)
+
+    unit = units(IFF.parent_ids, IFF.field)
+    if !(isempty(unit) || unit == "-")
+        printstyled(io, " [$unit]"; color=:yellow, bold=true)
```
Comment (orso82) [inline src/data.jl:913]: Yellow does not work on white background

File: src/data.jl [hunk 3]
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
...
+julia> IFF[end].value
+```
+"""
+function Base.findall(root_ids_arr::AbstractArray, target_fields::Union{Symbol,AbstractArray{Symbol},Regex}=r""; include_subfields::Bool=true)
+    target_fields == :all ? target_fields = r"" : target_fields
+
+    IFF_arr = Vector{IDS_Field_Finder}()
+    for root_ids in root_ids_arr
+        if root_ids isa Union{IDS,IDSvector}
+            append!(IFF_arr, Base.findall(root_ids, target_fields; include_subfields))
```
Comment (orso82) [inline src/data.jl:990]:
```suggestion
            append!(IFF_arr, findall(root_ids, target_fields; include_subfields))
```

File: src/data.jl [hunk 4]
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
...
+        end
+    end
+    return IFF_arr
+end
+
+function Base.findall(root_ids::Union{IDS,IDSvector}, target::Union{Symbol,AbstractArray{Symbol},Regex}=r""; include_subfields::Bool=true)
+    target == :all ? target = r"" : target
```
Comment (orso82) [inline src/data.jl:997]:
```suggestion
```

File: src/data.jl [hunk 5]
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
...
+- `target_fields::Union{Symbol, AbstractArray{Symbol}, Regex} = r""`: Fields to search for, specified by a single symbol, array of symbols, or regular expression.
+- `include_subfields::Bool = true`: If `true`, retrieves nested fields below the target field when found; if `false`, stops at the matching field.
+
+# Returns
+- `Vector{IDS_Field_Finder}`: A vector of `IDS_Field_Finder` structures, each containing details on a located field such as parent IDS, root IDS, field type, and full field path.
+
+# Example
+```julia-repl
+julia> findall(dd.equilibrium.time_slice[].global_quantities) # By default, it searches everything under given IDS
+julia> findall(dd.equilibrium.time_slice[].global_quantities, :all) # Same behavior
```
Comment (orso82) [inline src/data.jl:960]:
```suggestion
```

File: src/data.jl [hunk 6]
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
...
+    root_ids::Union{IDS,IDSvector} # Start point of the search
+    parent_ids::Union{IDS,IDSvector} # Parent IDS of target field
+    field::Symbol # Target field symbol
+    field_type::Type # Type of the field
+    field_path::String # Relative path from root_ids to the field
+
+    # Named constructor with keyword arguments for clear field assignment
+    IDS_Field_Finder(; root_ids, parent_ids, field, field_type, field_path) =
+        new(root_ids, parent_ids, field, field_type, field_path)
+end
```
Comment (orso82) [inline src/data.jl:880]:
```suggestion
Base.@kwdef struct IDS_Field_Finder
    root_ids::Union{IDS,IDSvector} # Start point of the search
    parent_ids::Union{IDS,IDSvector} # Parent IDS of target field
    field::Symbol # Target field symbol
    field_type::Type # Type of the field
    field_path::String # Relative path from root_ids to the field
end
```

**Review by orso82 — COMMENTED** [review]
this is great! Thank you @mgyoo86 

I have some minor comments

**Comment by orso82** [comment]
I'd like to avoid this syntax. It doesn't add anything and it conflicts with the idea of plotting a quantity named `:all`
```
findall(dd.equilibrium.time_slice[].global_quantities, :all)
```

**Comment by mgyoo86** [comment]
Thanks for the quick and thorough review! @orso82 
I reflected on all of your comments as follows:
- changed color from ":yellow" to ":blue"
- removed the ":all" keywords.

**Comment by orso82** [comment]
I have tested it and merged it. It may be good to write a regression test for this.

**Pull request merged**